### DISCUSSION TITLE
Handling buffer renames while switching buffers

### DISCUSSION THREAD

**Discussion by Stebalien** [minad/consult#979] [Ideas]
If a buffer is renamed after running `consult-buffer`, attempting to switch to it will open a new blank buffer with the original name. This is especially problematic in my EXWM setup, where previewing a buffer can change the buffer's name (e.g. previewing a slack window clears the unread messages count from the title).

Would it be possible to somehow record the original buffer name to buffer mapping?

**Comment by minad** [comment]
Buffer renaming is a well-known difficulty. The problem occurs also with Embark, when the minibuffer stays open while killing buffers, when buffers are renamed after a killing. I currently do not preview EXWM buffers, for this and various other reasons (focus issues). Unfortunately I don't have a simple solution for this problem. The current architecture relies quite strongly on the string identity of candidates. Maybe `consult-buffer` should first retrieve the actual buffer objects and attach them to the candidate strings, and then proceed from there. This change will then also affect Marginalia and Embark. It is certainly not a small fix.

**Reply by minad** [reply]
To create a plan for this: Fortunately all the buffers are obtained by `consult--buffer-query` with an `:as #'buffer-name` argument. Instead  of `buffer-name` we could use `(cons (buffer-name buffer) buffer))` and then instead attach the buffer as the `cdr` of the `multi-category`. This would be an extension of the `multi-category` protocol. Then `consult--buffer-state` and `marginalia-annotate-buffer` would have to handle both strings and buffer objects. Same with Embark.

This seems resolvable, but needs some coordination, also with Omar. I am currently a bit flooded, but hopefully I find time for this. Otherwise patches are also welcome.

**Reply by Stebalien** [reply]
Yeah, I figured this wasn't going to be easy. Unfortunately, completion does not preserve text properties so we can't just attach the buffer objects to the buffer names.

**Reply by Stebalien** [reply]
Ah, sorry, I should have reloaded before posting. I didn't realize there was a way to preserve the `cdr` through completion.

Unfortunately, I'll be down to one hand for the next couple of months so I won't be able to get much coding done. I'm also not sure if I understand the solution well enough to implement it without bugging you a lot.

**Reply by minad** [reply]
Actually, I got kind of triggered by this problem now. And I think it is easier than I had expected.

See the branch multi-category-buffer: https://github.com/minad/consult/tree/multi-category-buffer

Everything seems to work magically, since we can rely on the multi-category mechanism and get-buffer taking both strings and buffer objects (loose dynamic typing ftw!). So far I haven't tested Embark, but Marginalia and Consult seem to work.

**Reply by minad** [reply]
Let's also ping @oantolin for his opinion (and for potential coordination with Embark if necessary). Omar, do you think this will work with Embark if we attach the actual buffers as cdr of multi-category? See the aforementioned branch.

**Reply by minad** [reply]
Incidentally I just had a discussion if we can generalize multi-category to handle non-string values in the cdr. At least in limited scenarios (like buffers) we can make this work. See the dicussion https://github.com/minad/consult/discussions/977.

Thinking about Embark (still not tested yet), there we have to be more careful. Embark assumes strings for its actions and injects them. But it should be easy to work around this problem. Embark simply has to convert buffers to strings in its `multi-category` transformer (a little bit of buffer-specific code). @oantolin what do you think?

**Reply by minad** [reply]
Embark would only need a replacement of `embark--refine-multi-category` I think.

~~~ elisp
;; EDIT: See below for an improved version.
(defun embark--refine-multi-category (_type target)
  "Refine `multi-category' TARGET to its actual type."
  (pcase (get-text-property 0 'multi-category target)
    (`nil (cons 'general target))
    ;; PROBLEM/TODO: `buffer-name' returns nil if buffer is dead. 
    ;; Does Embark handle nil return values? 
    ;; Targets getting lost during refinement...
    ((and `(buffer . ,b) (guard (bufferp b)) (buffer-name b)))
    (x x)))
~~~

**Reply by minad** [reply]
@Stebalien If you could give this some testing, I would appreciate it. Ideally testing all the different ways to interact with buffers (in Consult and via Embark).

**Reply by Stebalien** [reply]
This definitely fixes my issue in Consult, but I don't use Embark so I can't easily test that.

**Reply by minad** [reply]
Okay, thanks. Let's wait for @oantolin's opinion then. I am testing with disabled Embark post hooks, keeping the minibuffer alive, and with disabled Marginalia cache: 

~~~ elisp
(setq marginalia--cache-size 0
      embark-post-action-hooks nil
      embark-quit-after-action nil)
~~~

Embark needs this updated refinement functionn:

~~~ elisp
(defun embark--refine-multi-category (_type target)
  "Refine `multi-category' TARGET to its actual type."
  (or (let ((mc (get-text-property 0 'multi-category target)))
        (cond
         ((and (eq (car mc) 'buffer) (buffer-live-p (cdr mc)))
          (cons 'buffer (buffer-name (cdr mc))))
         ((stringp (cdr mc)) mc)))
      (cons 'general target)))
~~~

Marginalia needs a more robust buffer annotator, which also handles dead buffers:

~~~ elisp
(defun marginalia-annotate-buffer (cand)
  "Annotate buffer CAND with modification status, file name and major mode."
  (when-let ((buffer (get-buffer cand)))
    (if (buffer-live-p buffer)
        (marginalia--fields
         ((marginalia--buffer-status buffer))
         ((marginalia--buffer-file buffer)
          :truncate -0.5 :face 'marginalia-file-name))
      (marginalia--fields ("(dead buffer)" :face 'error)))))
~~~

**Reply by oantolin** [reply]
This code looks good to me, but there is another option we could consider: for non-string target calling the action non-interactively with the target as an argument.

**Reply by minad** [reply]
@oantolin Interesting. But ultimately this strategy would be too dangerous, since it requires that every action supports both calling conventions. Note that Embark can act on buffer-strings in switch-to-buffer and buffer-objects in consult-buffer. Therefore I think converting back to strings in Embark during refinement is the better approach for now. Given that this is the first situation where we allow non-string candidates, we probably still have to figure out the best way and we may adjust Embark in the future.

**Reply by oantolin** [reply]
Yes, this is certainly a less invasive change. Let's convert back and forth between strings then (that's what I call the Unix philosophy 😛).

**Reply by minad** [reply]
Haha, yes. Embark pretty much follows the Unix philosophy with strings for everything like in Bash or Tcl. :-P

PR is here: https://github.com/oantolin/embark/pull/710

**Comment by minad** [comment]
All the patches have been merged (Consult, Marginalia, Embark). Closing.

**Reply by Stebalien** [reply]
Thank you!
